### PR TITLE
ROX-12504: Extract selector pacakge

### PR DESCRIPTION
**Pull request by fredrb**
## Description

This PR is a prefactor of two follow up changes in the resource listeners:
- Deployment stores exposing specific queries and updating deployment wrap objects
- Implementing deployment resolvers to remove re-sync

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

- CI

### CODE REVIEW

File: sensor/kubernetes/selector/selector.go [hunk 1]
```diff
@@ -1,50 +1,55 @@
...
-func createLabelsWithLen(labels map[string]string) labelWithLenImpl {
+// CreateLabelsWithLen create labels wrapper that respects the LabelsWithLen interface
+func CreateLabelsWithLen(labels map[string]string) LabelsWithLen {
 	return labelWithLenImpl{labels}
 }
 
-// selectorWrap holds a selector and information allowing for additional checks before matching
-type selectorWrap struct {
+// Wrap holds a selector and information allowing for additional checks before matching
+type Wrap struct {
```
Comment (lvalerom) [inline sensor/kubernetes/selector/selector.go:45]: [nit] I don't think we need to export this.

File: sensor/kubernetes/selector/selector.go [hunk 2]
```diff
@@ -66,31 +72,32 @@ func (d selectorDisjunction) Matches(labels labelsWithLen) bool {
...
 		sw.matchNil = true
 	}
 }
 
-// createSelector returns a SelectorWrapper for the given map of labels; matchNil determines whether
+// CreateSelector returns a SelectorWrapper for the given map of labels; matchNil determines whether
 // an empty set of labels matches everything or nothing.
-func createSelector(labelsMap map[string]string, opts ...selectorWrapOption) selectorWrap {
-	selWrapper := selectorWrap{matchNil: false}
+func CreateSelector(labelsMap map[string]string, opts ...WrapOption) Wrap {
```
Comment (lvalerom) [inline sensor/kubernetes/selector/selector.go:99]: [nit] I think we could return the `Selector` interface here.

**Review by lvalerom — APPROVED** [review]
Looks good to me. Just a couple of optional suggestions.

File: sensor/kubernetes/selector/selector.go (hunk 1 above)
Comment (fredrb) [reply sensor/kubernetes/selector/selector.go]: Done

File: sensor/kubernetes/selector/selector.go (hunk 2 above)
Comment (fredrb) [reply sensor/kubernetes/selector/selector.go]: Done

File: sensor/kubernetes/selector/selector.go [hunk 3]
```diff
@@ -66,31 +72,32 @@ func (d selectorDisjunction) Matches(labels labelsWithLen) bool {
...
-// createSelector returns a SelectorWrapper for the given map of labels; matchNil determines whether
+// CreateSelector returns a SelectorWrapper for the given map of labels; matchNil determines whether
 // an empty set of labels matches everything or nothing.
-func createSelector(labelsMap map[string]string, opts ...selectorWrapOption) selectorWrap {
-	selWrapper := selectorWrap{matchNil: false}
+func CreateSelector(labelsMap map[string]string, opts ...Options) Selector {
+	selWrapper := wrap{matchNil: false}
 
 	for _, opt := range opts {
 		opt(&selWrapper)
```
Comment (lvalerom) [inline sensor/kubernetes/selector/selector.go:103]:
```suggestion
	selWrapper := &wrap{matchNil: false}

	for _, opt := range opts {
		opt(selWrapper)
```

Comment (lvalerom) [inline sensor/kubernetes/selector/selector.go:103]:
```suggestion
	selWrapper := &wrap{matchNil: false}

	for _, opt := range opts {
		opt(selWrapper)
```

File: sensor/kubernetes/selector/selector_test.go [hunk 4]
```diff
@@ -64,70 +64,70 @@ func (m mockSelector) Matches(labels labels.Labels) bool {
 	return m.internalSelector.Matches(labels)
 }
 
-func (s *SelectorWrapperTestSuite) injectMockSelector(sw *selectorWrap) {
+func (s *SelectorWrapperTestSuite) injectMockSelector(sw wrap) {
```
Comment (lvalerom) [inline sensor/kubernetes/selector/selector_test.go:67]:
```suggestion
func (s *SelectorWrapperTestSuite) injectMockSelector(sw *wrap) {
```

File: sensor/kubernetes/selector/selector_test.go [hunk 5]
```diff
@@ -136,10 +136,12 @@ func (s *SelectorWrapperTestSuite) TestLabelMatching() {
 	for name, tt := range tests {
 		s.Run(name, func() {
 			s.hasMatchesBeenCalled = false
-			selectorWrap := createSelector(tt.givenSelectorLabels, tt.matchEmptySelector)
-			s.injectMockSelector(&selectorWrap)
+			selectorWrap := CreateSelector(tt.givenSelectorLabels, tt.matchEmptySelector)
+			wrapObj, ok := selectorWrap.(wrap)
```
Comment (lvalerom) [inline sensor/kubernetes/selector/selector_test.go:140]:
```suggestion
			wrapObj, ok := selectorWrap.(*wrap)
```

File: sensor/kubernetes/selector/selector_test.go [hunk 6]
```diff
@@ -199,16 +201,19 @@ func (s *SelectorWrapperTestSuite) TestLabelMatchingWithDisjunctions() {
 	for name, tt := range tests {
 		s.Run(name, func() {
 			s.hasMatchesBeenCalled = false
-			var selectorWrappers []selector
+			var selectorWrappers []Selector
 			for i, label := range tt.givenSelectorLabels {
-				newSelector := createSelector(label, tt.matchEmptySelector[i])
-				s.injectMockSelector(&newSelector)
+				newSelector := CreateSelector(label, tt.matchEmptySelector[i])
+				wrapObj, ok := newSelector.(wrap)
```
Comment (lvalerom) [inline sensor/kubernetes/selector/selector_test.go:207]:
```suggestion
				wrapObj, ok := newSelector.(*wrap)
```

**Review by lvalerom — CHANGES_REQUESTED** [review]
I think these changes are needed to make the tests pass.

**Review by lvalerom — APPROVED** [review]

**Pull request merged**